src/device_rx.c

```diff
@@ -0,0 +1,581 @@
+    char interface_name[ASTARTE_INTERFACE_NAME_MAX_SIZE] = { 0 };
+    int ret = snprintf(interface_name, ASTARTE_INTERFACE_NAME_MAX_SIZE, "%.*s", interface_name_len,
+        interface_name_start);
+    if ((ret < 0) || (ret >= ASTARTE_INTERFACE_NAME_MAX_SIZE)) {
+        ASTARTE_LOG_ERR("Error encoding interface name");
+        goto exit;
+    }
+
+    size_t path_len = topic_len - MQTT_BASE_TOPIC_LEN - strlen("/") - interface_name_len;
+    path = calloc(path_len + 1, sizeof(char));
+    if (!path) {
+        ASTARTE_LOG_ERR("Out of memory %s: %d", __FILE__, __LINE__);
+        goto exit;
+    }
+    memcpy(path, path_start, path_len);
+
+    on_data_message(device, interface_name, path, data, data_len);
+
+exit:
+    free(path);
+}
+
+/************************************************
+ *         Static functions definitions         *
+ ***********************************************/
+
+static void on_control_message(
+    astarte_device_handle_t device, const char *topic, const char *data, size_t data_len)
+{
+    if (strcmp(topic, device->control_consumer_prop_topic) == 0) {
+#if defined(CONFIG_ASTARTE_DEVICE_SDK_NVS)
+        on_purge_properties(device, data, data_len);
+#else
+        (void) device;
+        (void) data;
+        (void) data_len;
+#endif
+    } else {
+        ASTARTE_LOG_ERR("Received unrecognized control message: %s.", topic);
+    }
+}
+
+#if defined(CONFIG_ASTARTE_DEVICE_SDK_NVS)
+static void on_purge_properties(astarte_device_handle_t device, const char *data, size_t data_len)
+{
+    char *decomp_data = NULL;
+    dlist_t allow_list = dlist_init();
+
+    uLongf decomp_data_len = __builtin_bswap32(*(uint32_t *) data);
+
+    decomp_data = calloc(decomp_data_len + 1, sizeof(char));
+    if (!decomp_data) {
+        ASTARTE_LOG_ERR("Out of memory %s: %d", __FILE__, __LINE__);
+        goto exit;
+    }
+
+    if (decomp_data_len != 0) {
+        int uncompress_res = uncompress((char unsigned *) decomp_data, &decomp_data_len,
+            (char unsigned *) data + 4, data_len - 4);
+        if (uncompress_res != Z_OK) {
+            ASTARTE_LOG_ERR("Decompression error %d.", uncompress_res);
+            goto exit;
+        }
+    }
+
+    ASTARTE_LOG_DBG("Received purge properties: '%s'", decomp_data);
+
+    // Parse the received message and fill a list of properties
+    if (decomp_data_len != 0) {
+        char *property = strtok(decomp_data, ";");
+        if (!property) {
+            ASTARTE_LOG_ERR("Error parsing the purge property message %s.", decomp_data);
+            goto exit;
+        }
+        do {
+            astarte_result_t ares = dlist_append(&allow_list, property);
+            if (ares != ASTARTE_RESULT_OK) {
+                ASTARTE_LOG_ERR("Failure adding property in allow list.error: %s",
+                    astarte_result_to_name(ares));
+                goto exit;
+            }
+            property = strtok(NULL, ";");
+        } while (property);
+    }
```

Comment:
What do you think to check `decomp_data_len` only the the first time? 
```suggestion
if (decomp_data_len == 0) {
    goto exit;
  }

  int uncompress_res = uncompress((char unsigned*)decomp_data, &decomp_data_len,
                                  (char unsigned*)data + 4, data_len - 4);
  if (uncompress_res != Z_OK) {
    ASTARTE_LOG_ERR("Decompression error %d.", uncompress_res);
    goto exit;
  }

  ASTARTE_LOG_DBG("Received purge properties: '%s'", decomp_data);

  // Parse the received message and fill a list of properties
  char* property = strtok(decomp_data, ";");
  if (!property) {
    ASTARTE_LOG_ERR("Error parsing the purge property message %s.", decomp_data);
    goto exit;
  }
  do {
    astarte_result_t ares = dlist_append(&allow_list, property);
    if (ares != ASTARTE_RESULT_OK) {
      ASTARTE_LOG_ERR("Failure adding property in allow list.error: %s",
                      astarte_result_to_name(ares));
      goto exit;
    }
    property = strtok(NULL, ";");
  } while (property);
```